File: app/src/main/java/com/example/mike/mp3player/service/library/content/retriever/ContentRetriever.java
package com.example.mike.mp3player.service.library.content.retriever;

import android.support.v4.media.MediaBrowserCompat;

import com.example.mike.mp3player.commons.MediaItemType;
import com.example.mike.mp3player.service.library.ContentRequest;
import com.example.mike.mp3player.service.library.utils.IdGenerator;

import java.util.List;

public abstract class ContentRetriever {

    public abstract List<MediaBrowserCompat.MediaItem> getChildren(ContentRequest request);
    public abstract MediaItemType getType();


}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/content/retrievers/ContentRetriever.kt
package com.github.goldy1992.mp3player.service.library.content.retrievers

import androidx.media3.common.MediaItem
import com.github.goldy1992.mp3player.commons.MediaItemType

/**
 * Represents the fundamentals of a Content Retriever
 */
abstract class ContentRetriever {
    abstract fun getItems() : List<MediaItem>

    abstract fun getChildren(parentId : String) : List<MediaItem>

    /**
     * @return The type of MediaItem retrieved from the Content Retriever
     */
    abstract val type: MediaItemType
}